File: 0425-word-squares/0425-word-squares.cpp
struct Trie{
    Trie* arr[26];
    // vector<Trie*> arr;
    vector<const string *> startsWith;
    Trie(){
        for(int i=0; i<26; i++){
            arr[i] = NULL;
        }
        // arr.resize(26, NULL);
    }
};

class TrieDs{
    Trie *root;
public:
    TrieDs(vector<string> &arr){
        root = new Trie();
        for(int i=0; i<arr.size(); i++){
            insert(arr[i]);
        }
    }
    void insert(const string &key){
        Trie *cur=root;
        for(int i=0; i<key.size(); i++){
            int ind = key[i] - 'a';
            if(cur->arr[ind] == NULL){
                cur->arr[ind] = new Trie();
            }
            cur = cur->arr[ind];
            cur->startsWith.push_back(&key);
        }
    }

    vector<const string *> searchAllPrefix(const string &key){
        vector<string> res;
        Trie *cur=root;
        for(int i=0; i<key.size(); i++){
            int ind = key[i] - 'a';
            if(cur->arr[ind] == NULL){
                return {};
            }
            cur = cur->arr[ind];
        }
        return cur->startsWith;
    }
};

void dfs(TrieDs &ds, vector<vector<string>> &res, vector<string> &cur){
    if(cur.size() == cur[0].size()){
        res.push_back(cur);
        return;
    }
    
    string prefix = "";
    int ind = cur.size();
    for(const string &w:cur){
        prefix.push_back(w[ind]);
    }

    vector<const string *> wordList = ds.searchAllPrefix(prefix);
    
    for(const string *s: wordList){
        cur.push_back((*s));
        dfs(ds, res, cur);
        cur.pop_back();
    }

}
class Solution {
public:
vector<vector<string>> wordSquares(vector<string>& words) {
    vector<vector<string>> res;
    TrieDs ds(words);
    vector<string> cur;
    for(int i=0; i<words.size(); i++){
        cur.push_back(words[i]);
        dfs(ds, res, cur);
        cur.pop_back();
    }
    return res;
}
};

File: 0425-word-squares/0425-word-squares.py
def sss(l):
    n=len(l)
    m=len(l[0])
    d={"":l[:]}
    for i in range(n):
        for j in range(1,m+1):
            s=l[i][:j]
            if s not in d:
                d[s]=[]
            d[s].append(l[i])
    # print(d)
    ans=[]
    
    def ss(i,tt):
        t=tt[:]
        if i>=m:
            ans.append(t)
            return 
        s=t[i][:i]
        # x=0
        # for j in range(m):
        #     if t[j][:i]==s:
        #         x+=1 
         
        # print(i,t,s)
        if s not in d :
            return 
        
        else:
            for p in d[s]:
                t=tt[:]
                for j in range(i,m):
                    t[j] = t[j][:i]+p[j]+"."*(m-i-1)
                t[i]=p 
                ss(i+1,t)
           
    ss(0,["."*m]*m)           
    return ans

class Solution:
    def wordSquares(self, words: List[str]) -> List[List[str]]:
        return sss(words)
        